src/operator/optimizer_op-inl.h

```diff
@@ -460,6 +461,95 @@ inline void SGDMomUpdateRspRspRspImpl(const SGDMomParam& param,
                                  mom.data(), req, &out_blob);
 }
 
+inline bool SGDMomStorageType(const nnvm::NodeAttrs& attrs,
+                              const int dev_mask,
+                              DispatchMode* dispatch_mode,
+                              std::vector<int>* in_attrs,
+                              std::vector<int>* out_attrs) {
+  CHECK_EQ(in_attrs->size(), 3U);
+  CHECK_EQ(out_attrs->size(), 1U);
+  bool dispatched = false;
+  const NDArrayStorageType weight_stype = static_cast<NDArrayStorageType>(in_attrs->at(0));
+  const NDArrayStorageType grad_stype = static_cast<NDArrayStorageType>(in_attrs->at(1));
+  const NDArrayStorageType mom_stype = static_cast<NDArrayStorageType>(in_attrs->at(2));
+  if (!dispatched && common::ContainsOnlyStorage(*in_attrs, kDefaultStorage)) {
+    // dns, dns, dns -> dns
+    dispatched = storage_type_assign(out_attrs, kDefaultStorage,
+                                     dispatch_mode, DispatchMode::kFCompute);
+  }
+  if (!dispatched && weight_stype == kRowSparseStorage &&
+      grad_stype == kRowSparseStorage &&
+      (mom_stype == kRowSparseStorage || mom_stype == kDefaultStorage)) {
+    // rsp, rsp, rsp/dns -> rsp
+    dispatched = storage_type_assign(out_attrs, kRowSparseStorage,
+                                     dispatch_mode, DispatchMode::kFComputeEx);
+  }
+  if (!dispatched) {
+    dispatch_fallback(out_attrs, dispatch_mode);
+    LogStorageFallback(attrs, dev_mask, in_attrs, out_attrs);
+  }
+  return true;
+}
+
+template<int req>
+struct SGDMomStdDnsRspDnsKernel {
+  template<typename DType, typename IType, typename RType>
+  MSHADOW_XINLINE static void Map(int i, index_t row_length, DType* out_data,
+    DType* mom_data, const DType* weight_data, const IType* grad_idx,
+    const DType* grad_data, const RType* prefix_sum, const DType clip_gradient,
+    const DType momentum, const DType lr, const DType wd, const DType rescale_grad) {
+    const DType rate = lr * wd;
+    const bool non_zero = (i == 0) ? prefix_sum[0] > 0
+                                   : prefix_sum[i] > prefix_sum[i-1];
+
+    for (index_t j = 0; j < row_length; j++) {
+      const index_t data_i = i * row_length + j;
```

Comment:
`i * row_length ` can be cached and computed only once. 
Same for `(prefix_sum[i]-1) * row_length`